### PR TITLE
show client side validation for duplicate keyword entry

### LINKED ISSUE #1502: Notify user and prevent saving when duplicate keyword has been entered
Our keywords map to DOI subjects, which has a uniqueness requirement in the DataCite DOI API. See https://github.com/datacite/schema/blob/master/source/json/kernel-4.3/datacite_4.3_schema.json#L298.

If a user enters duplicate keywords, the DOI is not created, though a DOI displays in H2 as if it has been.

The request here is to identify duplicates via a string match against other entries on the current form so that users can delete one to prevent the DOI minting from running into trouble. Users should be able to save as draft, submit for approval, or deposit when there are duplicate keywords. Display error message and red box around field similar to other validation errors.

image

### PR DESCRIPTION
**Pull request by peetucket**
## Why was this change made? 🤔

Fixes #1502 - adds client side duplicate keyword validation

Notes:
- for simplicity of implementation, I reused the existing "invalid selection" that typically shows for blank keywords and expanded the error message from "Keyword must be filled in" to "Keyword must be filled in and be unique" so there is a single client side error message that will show in either case
- again for simplicity, the error shows next to the last text box that lost focus when a duplicate was detected, so it may be possible to get into some situations where the error shows on the wrong text box, though i think this simple method should work fine in most cases

Example:

H2-keywords

## How was this change tested? 🤨

Localhost

### CODE REVIEW

File: app/components/works/keywords_row_component.html.erb [hunk 1]
```diff
@@ -1,15 +1,16 @@
-<div class="mb-3 row plain-container g-0">
+<div class="mb-3 row plain-container g-0 keyword-row">
   <div class="col-sm-6">
     <%= form.label :label, 'Keyword', class: 'col-form-label sr-only' %>
-    <div data-controller="autocomplete autocomplete-edit" data-autocomplete-url-value="/autocomplete" data-autocomplete-min-length-value="3" class="dropdown">
-      <%= form.text_field :label, { class: "form-control#{ 'is-invalid' if error?}", required: true, 'data-autocomplete-target': 'input', 'data-autocomplete-edit-target': 'input' } %>
+    <div data-controller="keywords autocomplete autocomplete-edit" data-autocomplete-url-value="/autocomplete" data-autocomplete-min-length-value="3" class="dropdown">
+      <%= form.text_field :label, { class: "form-control#{ 'is-invalid' if error?}", required: true, 'data-autocomplete-target': 'input', 'data-autocomplete-edit-target': 'input',
+                'data-action': 'blur->keywords#checkForDuplicates' } %>
       <!-- This input is populated by autocomplete controller.-->
```
Comment (peetucket) [inline app/components/works/keywords_row_component.html.erb:7]: just added a new action to the keyword text box to check for duplicates on losing focus

File: app/javascript/controllers/keywords_controller.js [hunk 2]
```diff
@@ -23,16 +23,23 @@ export default class extends Controller {
     this.containerTarget.classList.remove('is-invalid')
   }
 
-  checkForEnter(event) {
-    if (event.key === 'Enter') {
-      event.preventDefault() // prevent form submission
-      this.add(event);
```
Comment (peetucket) [inline app/javascript/controllers/keywords_controller.js:29]: i spotted this similar method to what we just implemented for sunetids, but could not find any references to this particular "checkForEnter()" method anywhere in the code, so I removed it

File: app/javascript/controllers/keywords_controller.js [hunk 3]
```diff
@@ -23,16 +23,23 @@ export default class extends Controller {
...
+  checkForDuplicates(event) {
+    const input = event.target.closest('input[type="text"]')
+    const keywords = Array.from(document.querySelectorAll('.keyword-row:not([style*="display: none"]) input[type="text"]'), input => input.value)
```
Comment (peetucket) [inline app/javascript/controllers/keywords_controller.js:28]: this is the gnarly selector -- basically look for any non-hidden `.keyword-row` elements that have a text box, and map out their values into an array

File: app/javascript/controllers/keywords_controller.js [hunk 4]
```diff
@@ -23,16 +23,23 @@ export default class extends Controller {
     this.containerTarget.classList.remove('is-invalid')
   }
 
-  checkForEnter(event) {
-    if (event.key === 'Enter') {
-      event.preventDefault() // prevent form submission
-      this.add(event);
+  checkForDuplicates(event) {
+    const input = event.target.closest('input[type="text"]')
```
Comment (jcoyne) [inline app/javascript/controllers/keywords_controller.js:27]: Could you use a stimulus target for this?

File: app/javascript/controllers/keywords_controller.js (hunk 3 above)
Comment (jcoyne) [reply app/javascript/controllers/keywords_controller.js:28]: How about extracting this into a hasDuplicateKewords method?  This looks pretty similar to what we have in https://github.com/sul-dlss/argo/blob/6139bc519f915c25cd8e55eccd36a6708b22ef49/app/javascript/controllers/registration_item_row_controller.js#L94-L100

**Review by jcoyne — CHANGES_REQUESTED** [review]

File: app/javascript/controllers/keywords_controller.js (hunk 4 above)
Comment (peetucket) [reply app/javascript/controllers/keywords_controller.js]: Yup - done

**Review by jcoyne — APPROVED** [review]

**Pull request merged**